### ISSUE TITLE
Add support for .NET Core

### ISSUE THREAD

**Issue by ghost** [approvals/ApprovalTests.Net#162]
Self-explanatory, but would be very nice to have.

**Comment by droyad** [comment]
I had a detailed go at this, and came up empty. The main issue I had was that I could not find a way to get the current `StackTrace`. This meant that the discovery of Attributes is not possible. 

Also without the `StackTrace`, the code can't get the source filename using `StackFrame.GetFileName()`, which is relied upon by the naming.

One work around would be to pass the calling method into Verify. Another is to parse the `Environment.StackTrace` string.

Alternatively the overload needed (`StackTrace(bool)`) will be in `netstandard2.0`, which is expected in Q4/Q1.

**Comment by isidore** [comment]
I'd be interested in pairing on this. should we wait until 2.0 ?

Llewellyn Falco
www.approvaltests.com

**Comment by fdmrun** [comment]
If a valid workaround is to just pass the calling method into Verify, then CallerMemberNameAttribute will do the job (https://msdn.microsoft.com/en-us/library/system.runtime.compilerservices.callermembernameattribute(v=vs.110).aspx).

**Comment by AndreyAkinshin** [comment]
What the current status of the issue? I really want to use ApprovalTests.Net in one of my .NET Core projects.

**Comment by isidore** [comment]
@fdmrun thanks for letting me know about CallerMemberNameAttribute. It's not enough for this, but I'm sure I'll use it in the future.

@AndreyAkinshin unsure. I think we are still waiting on netstandard2.0 but I'd be happy to explore this with you via skype if you have time this weekend?

**Comment by AndreyAkinshin** [comment]
> I'd be happy to explore this with you via skype if you have time this weekend?

@isidore, Ok, send me your skype login to e-mail (can be found in my profile).

**Comment by isidore** [comment]
Looks like stacktrace is still an issue as of this moment. 
I'm not currently working with .netcore; so can one of you ping me when this looks possible again?

thanks

**Comment by mscottreed** [comment]
[CallerMemberName] kinda makes stack tracing unnecessary.  I created a Approvals lite for .NET Core, and I have used it in two projects already.

**Comment by jamesrcounts** [comment]
@mscottreed might have meant [CallerFilePathAttribute]

**Comment by mscottreed** [comment]
Both actually

**Comment by EamonNerbonne** [comment]
Stack walking seems not to find the appropriate location in non-core (e.g. net46-targetting) assemblies that were compiled with the new .net core tooling.

The stack-trace walking code has always been flaky for me - if your build process ever moves around files, things can get tricky easily; and then theres the need to add [MethodImpl(MethodImplOptions.NoInlining)] all over the place.

I [currently use](https://github.com/EamonNerbonne/ExpressionToCode/blob/prep-dotnet-core/ExpressionToCodeTest/ApiStabilityTest.cs#L19) this small helper to avoid all (at least, I hope?) stack-walking code in approvals.net:

```C#
        class SaneNamer : IApprovalNamer
        {
            public string SourcePath { get; set; }
            public string Name { get; set; }
        }

        public static void Verify(string text, object IGNORE_PAST_THIS = null, [CallerFilePath] string filepath = null, [CallerMemberName] string membername = null)
        {
            var writer = WriterFactory.CreateTextWriter(text);
            var filename = Path.GetFileNameWithoutExtension(filepath);
            var filedir = Path.GetDirectoryName(filepath);
            var namer = new SaneNamer {Name = filename + "." + membername, SourcePath = filedir};
            var reporter = new DiffReporter();
            Approvals.Verify(writer, namer, reporter);
            Approver.Verify(new FileApprover(writer, namer, true), reporter);
        }
```

Personally, I'd like an API without stackwalking as the default.  Not only is that easier to use (no MethodImpl to forget, no deployment/postbuild interactions), you could also more easily stick it in a loop.  The current api makes all that very difficult because it's quite unclear how what aspects of the configuration are injected.  E.g. I love use approvals for refactoring, and often it's handy to split large test runs by some discriminator (e.g. language, client, etc.) to keep things manageable in size and make em easier to interpret.

**Comment by jamesrcounts** [comment]
Agree with @EamonNerbonne stack walking should be retired.  In modern versions of .NET it doesn't seem necessary anymore (although it might have been the only solution in the past).  I'd be in favor of dropping support for old versions of .NET or making a new library to support modern .NET (and of course a Core version would be just that).

**Comment by droyad** [comment]
I've created [Assent](https://github.com/droyad/Assent) for our .NET Core approval needs. It's nowhere are comprehensive as ApprovalTests though.

**Comment by SeanFeldman** [comment]
What's the current state of this issue? In the original comments there was mentioning of netstandard2.0.
With netcore2.0 out (which which I assume people would run the tests) this should be doable.

**Comment by niemyjski** [comment]
I'm also interested in this.

**Comment by isidore** [comment]
@SimonCropp & I are working on this, it's taking a lot of yak shaving...

**Comment by niemyjski** [comment]
@isidore Are you working in any specific branches? It shouldn't take that much to port to .net standard 2.0

**Comment by SeanFeldman** [comment]
>It shouldn't take that much to port to .net standard 2.0

@niemyjski assumptions... our industry is full of those. Avoid adding more 😉

**Comment by niemyjski** [comment]
I’ve ported oauth2 lib, exceptionless net and foundatio libs in less than a week... some cases took me 2-3 hours.. I’m 4-5 days in and I almost have exceptionless server fully converted to asp.net core from all legacy so I really don’t think it would take much.......

**Comment by SeanFeldman** [comment]
https://upload.wikimedia.org/wikipedia/en/thumb/e/ea/Superman_shield.svg/1280px-Superman_shield.svg.png

**Comment by droyad** [comment]
I had a crack a few months ago, it looked like much muchness would be required

**Comment by PaulDMendoza** [comment]
@isidore Any updates on this? I'm trying to convert a test project to .NET Core 2.0 from .NET 4.5 and just realized this thread is open still.

**Comment by isidore** [comment]
we are still working on it. it's more complicated than expected.

Llewellyn Falco
www.approvaltests.com

**Comment by PaulDMendoza** [comment]
Any updates on this getting fixed?

**Comment by niemyjski** [comment]
Were we able to see if there are any blockers upgrading yet?

**Comment by dionfoster** [comment]
Hey guys, just wanted to try find out the progress of this.  Are there blockers from external dependencies?  What problems are there in getting it to netcore?  Are there fundamental differences or lack of support from the framework that Approvals relies on?

**Comment by PaulDMendoza** [comment]
I'm also still interested. Any updates?

**Comment by LeonidEfremov** [comment]
C'mon guys, we really need this update =))

**Comment by isidore** [comment]
Just an update.
We've been working on this, ApprovalTests has lots of dependencies on many different and incompatible technologies. I've been teasing it apart with @SimonCropp but if people are wanting it sooner I'd suggest setting up time to pair with me. (I'm not currently working in .net core, so pairing is only realistic way I make time to do this)

**Comment by niemyjski** [comment]
Just let me know when you have time to meet up and I'll make some time :).

Thanks
-Blake Niemyjski

**Comment by bhugot** [comment]
done something in #214

**Comment by isidore** [comment]
just an update. @SimonCropp  & I have separated a bunch of the things from approvaltests.

Leaving just
    <Reference Include="PresentationCore" Pack="False" />
    <Reference Include="PresentationFramework" Pack="False" />
    <Reference Include="System.Management" Pack="False" />
    <Reference Include="WindowsBase" Pack="False" />
&
    <Reference Include="System" Pack="False" />
    <Reference Include="System.Configuration" Pack="False" />

preventing us from a core compliant package. It's extremely slow work, but we are doing it and progress is being made.

**Comment by kl1mm** [comment]
are you getting forward with the port to .NetStandard? Would be amazing to have this at nuget

**Comment by JoshSchreuder** [comment]
Just tried out 3.0.18 which has `netstandard2.0` support and it seems to be working for my `netcoreapp2.1` test project running on Windows :)

**Comment by niemyjski** [comment]
I'll try it out, I seen it has a dependency on the registry.. Any specifics on this, I'm not sure that will work xplat.

**Comment by SimonCropp** [comment]
@niemyjski the registry ref is only to enable the WindowsRegistryAssert functionality, which you would not be going on xplat

closing this since netcore is now supported

**Comment by niemyjski** [comment]
Could this be added in a windows only specific package so we don't have to bring this in cross plat

**Comment by SimonCropp** [comment]
In this case i dont se the ROI. What exactly is your concern?

**Comment by benbelow** [comment]
@SimonCropp Do you expect this to be cross platform compatible then? I'm having some issues with `Alphaleonis.Win32.Filesystem.Path` running on an Ubuntu CI server
